sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/RowConverter.scala

```diff
@@ -3,246 +3,224 @@
+        elementType match {
+            case _: StructType => rowTypeRouterToJsonArray(data)
+            case subArray: ArrayType => convertSparkArrayToArrayNode(subArray.elementType, containsNull, data.asInstanceOf[Seq[_]])
+            case _ => convertSparkDataTypeToJsonNode(elementType, data)
+        }
     }
-  }
 
-  private def arrayTypeToJsonArray(elementType: DataType, data: Seq[Any], isInternalRow: Boolean): ArrayNode = {
-    val arrayNode = objectMapper.createArrayNode()
+    private def rowTypeRouterToJsonArray(element: Any) : ObjectNode = {
+        element match {
+            case e: Row => fromRowToObjectNode(e)
+            case _ => throw new Exception(s"Cannot cast $element into a Json value. Struct $element has no matching Json value.")
+        }
+    }
 
-    elementType match {
-      case subDocuments: StructType => data.foreach(x => arrayNode.add(rowTypeRouterToJsonArray(x, subDocuments)))
-      case subArray: ArrayType => data.foreach(x => arrayNode.add(arrayTypeRouterToJsonArray(subArray.elementType, x, isInternalRow)))
-      case _ => data.foreach(x => arrayNode.add(convertToJson(x, elementType, isInternalRow)))
+    private def convertStructToSparkDataType(schema: StructType, objectNode: ObjectNode) : Seq[Any] =
+        schema.fields.map {
+            case StructField(name, dataType, _, _) =>
+            Option(objectNode.get(name)).map(convertToSparkDataType(dataType, _)).orNull
+        }
+
+    // scalastyle:off
+    private def convertToSparkDataType(dataType: DataType, value: JsonNode): Any = (value, dataType) match {
+        case (_ : NullNode, _) | (_, _ : NullType) => null
+        case (jsonNode: ObjectNode, struct: StructType) =>
+            fromObjectNodeToRow(struct, jsonNode)
+        case (jsonNode: ObjectNode, map: MapType) =>
+            jsonNode.fields().asScala
+                .map(element => (
+                    element.getKey,
+                    convertToSparkDataType(map.valueType, element.getValue)))
+        case (arrayNode: ArrayNode, array: ArrayType) =>
+            arrayNode.elements().asScala.map(convertToSparkDataType(array.elementType, _)).toArray
+        case (binaryNode: BinaryNode, _: BinaryType) =>
+            binaryNode.binaryValue()
+        case (arrayNode: ArrayNode, _: BinaryType) =>
+            // Assuming the array is of bytes
+            objectMapper.convertValue(arrayNode, classOf[Array[Byte]])
+        case (_, _: BooleanType) => value.asBoolean()
+        case (_, _: StringType) => value.asText()
+        case (_, _: DateType) => toDate(value)
+        case (_, _: TimestampType) => toTimestamp(value)
+        case (isJsonNumber(), DoubleType) => value.asDouble()
+        case (isJsonNumber(), DecimalType()) => value.decimalValue()
+        case (isJsonNumber(), FloatType) => value.asDouble()
+        case (isJsonNumber(), LongType) => value.asLong()
```

Comment:
For these I think we also need a similar translation of text --> Float, Text -->Double etc. like for Date and Timestamp

Reply:
Done 😄